### PR TITLE
[BC-breaking] Use ScatterGatherKernel for scatter_reduce (CPU-only)

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74608
* #74607
* #74606
* __->__ #74226

Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce)`

- Add new reduction options in ScatterGatherKernel.cpp and update `scatter_reduce` to call into the cpu kernel for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Migrate `test/test_torch.py:test_scatter_reduce` to `test/test_scatter_gather_ops.py`

Differential Revision: [D35222842](https://our.internmc.facebook.com/intern/diff/D35222842)

### CODE REVIEW

File: aten/src/ATen/native/cuda/ScatterGatherKernel.cu [hunk 1]
```diff
@@ -244,6 +262,66 @@ struct cuda_scatter_gather_base_kernel {
...
+    auto self_dim_size = ensure_nonempty_size(self, dim);
+
+    auto src_dim_stride = ensure_nonempty_stride(src, dim);
+    auto src_dim_size = ensure_nonempty_size(src, dim);
+
+    auto index_size = self_dim_size;
+    auto index_stride = self_dim_stride;
+
+
+    AT_DISPATCH_FLOATING_TYPES_AND2(
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cuda/ScatterGatherKernel.cu:311]: If the only difference between this kernel and cuda_scatter_gather_base_kernel is 
`AT_DISPATCH_FLOATING_TYPES_AND2` vs `AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3` it might be easier to explicitly check for a specific scalar type and error out in a unified kernel than copy-pasting the code and just changing this line.

The error message to throw can be found in the [dispatch macro definitions](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Dispatch.h#L729).

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 2]
```diff
@@ -106,10 +106,9 @@
     ("aten::_scatter_reduce", datetime.date(2022, 1, 31)),
     ("aten::native_multi_head_self_attention", datetime.date(9999, 1, 1)),
     ("aten::_native_multi_head_self_attention", datetime.date(9999, 1, 1)),
-    ("aten::scatter_reduce.two", datetime.date(2022, 3, 15)),
     ("aten::grid_sampler_3d_backward", datetime.date(9999, 1, 1)),
     ("aten::_transform_bias_rescale_qkv", datetime.date(9999, 1, 1)),
-    ("aten::_scatter_reduce.two", datetime.date(9999, 1, 1)),
+    ("aten::scatter_reduce.two", datetime.date(9999, 1, 1)),
```
Comment (cpuhrsch) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:111]: Why this change?

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -15680,16 +15686,9 @@ def ref_pairwise_distance(input1, input2):
     ),
     OpInfo(
         'scatter_reduce',
-        dtypes=all_types_and(torch.float16, torch.bfloat16),
```
Comment (cpuhrsch) [inline torch/testing/_internal/common_methods_invocations.py:15683]: How come this was working beforehand?

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Looks pretty good, but I'd try to avoid the code duplication of copy-pasting `cpu_scatter_gather_base_kernel` to deal with the different set of dtypes.

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py (hunk 2 above)
Comment (mikaylagawarecki) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py:111]: Previously this was here when `_scatter_reduce` was moved to  `scatter_reduce`. Now I changed the signature of `scatter_reduce` so I assumed adding it to this dict was necessary

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (mikaylagawarecki) [reply torch/testing/_internal/common_methods_invocations.py:15683]: Previously this was enabled for cpu only, I sent a longer discussion regarding this offline

File: test/test_torch.py [hunk 4]
```diff
@@ -3297,6 +3297,56 @@ def test_scatter_add_bool(self, device):
                                             [False, True, False, True, False],
                                             [True, False, True, False, True]], device=device))
 
+    # FIXME: move to test_scatter_gather_ops.py
```
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply test/test_torch.py:3300]: Sorry about that, previously this test was in the class `TestTorch` which does not allow parameterization with @dtypes because it is not instantiated with `instantiate_device_type_tests(TestTorch, globals())`. Moved this here into the class `TestTorchDeviceType` which allows parameterization by @dtypes. Is there another way around this?

File: aten/src/ATen/native/cpu/ScatterGatherKernel.cpp [hunk 5]
```diff
@@ -333,6 +527,8 @@ void scatter_scalar_reduce_cpu_kernel(const Tensor& self, const int64_t dim, con
     cpu_scatter_gather_base_kernel<>()(self, dim, index, value,
                                        "scatter_scalar_reduce_multiply_", reduce_multiply);
     break;
+  default:
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:530]: Why did you add this?

File: test/test_scatter_gather_ops.py [hunk 6]
```diff
@@ -158,6 +179,46 @@ def test_scatter_add_mult_index_base(self, device, dtype):
         self.assertEqual(res0[0, :], m * torch.ones(n, device=device, dtype=dtype), atol=0, rtol=0)
         self.assertEqual(res1[:, 0], n * torch.ones(m, device=device, dtype=dtype), atol=0, rtol=0)
 
+    @onlyCPU
+    @toleranceOverride({torch.float16: tol(atol=1e-2, rtol=0),
```
Comment (cpuhrsch) [inline test/test_scatter_gather_ops.py:183]: How did you determine these tolerances?

File: aten/src/ATen/native/cpu/ScatterGatherKernel.cpp (hunk 5 above)
Comment (mikaylagawarecki) [reply aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:530]: Without this the build on CI will fail. I think it's due to this 

> Using the -Werror compiler flag, a switch statement over a value of an enum type without a default label will fail to compile if any enumerator of the enum doesn’t have a corresponding case. This is sometimes called an exhaustive or defaultless switch statement.

File: test/test_scatter_gather_ops.py (hunk 6 above)
Comment (mikaylagawarecki) [reply test/test_scatter_gather_ops.py:183]: For float16 this is copied from [the toleranceOverride decorator on test_scatter__reductions in this file](https://github.com/pytorch/pytorch/blob/22a3a17ea0d61660b53f6642fe22529a7e4e2649/test/test_scatter_gather_ops.py#L135). I copied this for bfloat16 but I'm not sure what tolerance should be used. How does one normally determine this

**Review by cpuhrsch — APPROVED** [review]
Looks great, just added two small comments

File: test/test_scatter_gather_ops.py (hunk 6 above)
Comment (cpuhrsch) [reply test/test_scatter_gather_ops.py:183]: There are default tolerances set, I'd first try running this without an override

Comment (mikaylagawarecki) [reply test/test_scatter_gather_ops.py]: Removed these decorators, thanks!

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35222842).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35222842).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35222842).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35222842).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35222842).
